File: lang/cs/Org.Apache.REEF.Wake/Time/Runtime/RuntimeClock.cs
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Implementations.InjectionPlan;
using Org.Apache.REEF.Utilities.Diagnostics;
using Org.Apache.REEF.Utilities.Logging;
using Org.Apache.REEF.Wake.RX.Impl;
using Org.Apache.REEF.Wake.Time.Event;
using Org.Apache.REEF.Wake.Time.Runtime.Event;

namespace Org.Apache.REEF.Wake.Time.Runtime
{
    public class RuntimeClock : IClock
    {
        private static readonly Logger LOGGER = Logger.GetLogger(typeof(RuntimeClock));

        private readonly ITimer _timer;
        private readonly PubSubSubject<Time> _handlers;
        private readonly ISet<Time> _schedule;

        private readonly IInjectionFuture<ISet<IObserver<StartTime>>> _startHandler;
        private readonly IInjectionFuture<ISet<IObserver<StopTime>>> _stopHandler;
        private IInjectionFuture<ISet<IObserver<RuntimeStart>>> _runtimeStartHandler;
        private IInjectionFuture<ISet<IObserver<RuntimeStop>>> _runtimeStopHandler;
        private readonly IInjectionFuture<ISet<IObserver<IdleClock>>> _idleHandler;

        private bool _disposed;
            
        /// <summary>
        /// Create a new RuntimeClock with injectable IObservers
        /// </summary>
        /// <param name="timer">The runtime clock timer</param>
        /// <param name="startHandler">The start handler</param>
        /// <param name="stopHandler">The stop handler</param>
        /// <param name="runtimeStartHandler">The runtime start handler</param>
        /// <param name="runtimeStopHandler">The runtime stop handler</param>
        /// <param name="idleHandler">The idle handler</param>
        [Inject]
        internal RuntimeClock(
            ITimer timer,
            [Parameter(typeof(StartHandler))] IInjectionFuture<ISet<IObserver<StartTime>>> startHandler, 
            [Parameter(typeof(StopHandler))] IInjectionFuture<ISet<IObserver<StopTime>>> stopHandler,
            [Parameter(typeof(RuntimeStartHandler))] IInjectionFuture<ISet<IObserver<RuntimeStart>>> runtimeStartHandler,
            [Parameter(typeof(RuntimeStopHandler))] IInjectionFuture<ISet<IObserver<RuntimeStop>>> runtimeStopHandler,
            [Parameter(typeof(IdleHandler))] IInjectionFuture<ISet<IObserver<IdleClock>>> idleHandler)
        {
            _timer = timer;
            _schedule = new SortedSet<Time>();
            _handlers = new PubSubSubject<Time>();

            _startHandler = startHandler;
            _stopHandler = stopHandler;
            _runtimeStartHandler = runtimeStartHandler;
            _runtimeStopHandler = runtimeStopHandler;
            _idleHandler = idleHandler;
        }

        public IInjectionFuture<ISet<IObserver<RuntimeStart>>> InjectedRuntimeStartHandler
        {
            get { return _runtimeStartHandler; }
            set { _runtimeStartHandler = value; }
        }

        public IInjectionFuture<ISet<IObserver<RuntimeStop>>> InjectedRuntimeStopHandler
        {
            get { return _runtimeStopHandler; }
            set { _runtimeStopHandler = value; }
        }

        /// <summary>
        /// Schedule a TimerEvent at the given future offset
        /// </summary>
        /// <param name="offset">The offset in the future to schedule the alarm</param>
        /// <param name="handler">The IObserver to to be called</param>
        public override void ScheduleAlarm(long offset, IObserver<Alarm> handler)
        {
            if (_disposed)
            {
                return;
            }
            if (handler == null)
            {
                Exceptions.Throw(new ArgumentNullException("handler"), LOGGER);
            }

            lock (_schedule)
            {
                _schedule.Add(new ClientAlarm(_timer.CurrentTime + offset, handler));
                Monitor.PulseAll(_schedule);
            }
        }

        /// <summary>
        /// Clock is idle if it has no future alarms set
        /// </summary>
        /// <returns>True if no future alarms are set, otherwise false</returns>
        public override bool IsIdle()
        {
            lock (_schedule)
            {
                return _schedule.Count == 0;
            }
        }

        /// <summary>
        /// Dispose of the clock and all scheduled alarms
        /// </summary>
        public override void Dispose()
        {
            lock (_schedule)
            {
                _schedule.Clear();
                _schedule.Add(new StopTime(_timer.CurrentTime));
                Monitor.PulseAll(_schedule);
                _disposed = true;
            }
        }

        /// <summary>
        /// Register the IObserver for the particular Time event.
        /// </summary>
        /// <param name="observer">The handler to register</param>
        public void RegisterObserver<U>(IObserver<U> observer) where U : Time
        {
            if (_disposed)
            {
                return;
            }

            _handlers.Subscribe(observer);
        }

        /// <summary>
        /// Start the RuntimeClock.
        /// Clock will continue to run and handle events until it has been disposed.
        /// </summary>
        public void Run()
        {
            SubscribeHandlers();
            _handlers.OnNext(new RuntimeStart(_timer.CurrentTime));
            _handlers.OnNext(new StartTime(_timer.CurrentTime));

            while (true)
            {
                lock (_schedule)
                {
                    if (IsIdle())
                    {
                        _handlers.OnNext(new IdleClock(_timer.CurrentTime));
                    }
                    
                    // Blocks and releases lock until it receives the next event
                    Time alarm = GetNextEvent();
                    ProcessEvent(alarm);

                    if (alarm is StopTime)
                    {
                        break;
                    }
                }
            }
            _handlers.OnNext(new RuntimeStop(_timer.CurrentTime));
        }

        /// <summary>
        /// Register the event handlers
        /// </summary>
        private void SubscribeHandlers()
        {
            Subscribe(_startHandler.Get());
            Subscribe(_stopHandler.Get());
            Subscribe(_runtimeStartHandler.Get());
            Subscribe(_runtimeStopHandler.Get());
            Subscribe(_idleHandler.Get());
        }

        /// <summary>
        /// Subscribe a set of IObservers for a particular Time event
        /// </summary>
        /// <param name="observers">The set of observers to subscribe</param>
        private void Subscribe<U>(ISet<IObserver<U>> observers) where U : Time
        {
            foreach (IObserver<U> observer in observers)
            {
                _handlers.Subscribe(observer);
            }
        }
            
        /// <summary>
        /// Wait until the first scheduled alarm is ready to be handled
        /// Assumes that we have a lock on the _schedule SortedSet
        /// </summary>
        private Time GetNextEvent()
        {
            // Wait for an alarm to be scheduled on the condition variable Count
            while (_schedule.Count == 0)
            {
                Monitor.Wait(_schedule);
            }

            // Once the alarm is scheduled, wait for the prescribed amount of time.
            // If a new alarm is scheduled with a shorter duration, Wait will preempt
            // and duration will update to reflect the new alarm's timestamp
            for (long duration = _timer.GetDuration(_schedule.First().TimeStamp);
                 duration > 0;
                 duration = _timer.GetDuration(_schedule.First().TimeStamp))
            {
                Monitor.Wait(_schedule, TimeSpan.FromMilliseconds(duration));
            }

            Time time = _schedule.First();
            _schedule.Remove(time);
            return time;
        }

        /// <summary>
        /// Process the next Time event. 
        /// </summary>
        /// <param name="time">The Time event to handle</param>
        private void ProcessEvent(Time time)
        {
            if (time is Alarm)
            {
                Alarm alarm = (Alarm)time;
                alarm.Handle();
            }
            else
            {
                _handlers.OnNext(time);
            }
        }
    }
}

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/RuntimeClock.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.wake.time.runtime;

import org.apache.reef.tang.InjectionFuture;
import org.apache.reef.tang.annotations.Parameter;
import org.apache.reef.wake.EventHandler;
import org.apache.reef.wake.impl.PubSubEventHandler;
import org.apache.reef.wake.time.Clock;
import org.apache.reef.wake.time.Time;
import org.apache.reef.wake.time.event.Alarm;
import org.apache.reef.wake.time.event.StartTime;
import org.apache.reef.wake.time.event.StopTime;
import org.apache.reef.wake.time.runtime.event.*;

import javax.inject.Inject;
import java.util.Set;
import java.util.TreeSet;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Default implementation of clock.
 *
 * After invoking `RuntimeStart` and `StartTime` events initially,
 * this invokes scheduled events on time. If there is no scheduled event,
 * `IdleClock` event is invoked.
 */
public final class RuntimeClock implements Clock {

  private static final Logger LOG = Logger.getLogger(RuntimeClock.class.getName());
  private static final String CLASS_NAME = RuntimeClock.class.getCanonicalName();

  /**
   * Injectable source of current time information.
   * Usually an instance of RealTimer that wraps the system clock.
   */
  private final Timer timer;

  /**
   * An ordered set of timed objects, in ascending order of their timestamps.
   * It also serves as the main synchronization monitor for the class.
   */
  private final TreeSet<Time> schedule = new TreeSet<>();

  /** Event handlers - populated with the injectable parameters provided to the RuntimeClock constructor. */
  private final PubSubEventHandler<Time> handlers = new PubSubEventHandler<>();

  private final InjectionFuture<Set<EventHandler<StartTime>>> startHandler;
  private final InjectionFuture<Set<EventHandler<StopTime>>> stopHandler;
  private final InjectionFuture<Set<EventHandler<RuntimeStart>>> runtimeStartHandler;
  private final InjectionFuture<Set<EventHandler<RuntimeStop>>> runtimeStopHandler;
  private final InjectionFuture<Set<EventHandler<IdleClock>>> idleHandler;

  /**
   * Timestamp of the last client alarm in the schedule.
   * We use it to schedule a graceful shutdown event immediately after all client alarms.
   */
  private long lastClientAlarm = 0;

  /**
   * Number of client alarms in the schedule.
   * We need it to determine whether event loop is idle (i.e. has no client alarms scheduled)
   */
  private int numClientAlarms = 0;

  /** Set to true when the clock is closed. */
  private boolean isClosed = false;

  /** Exception that caused the clock to stop. */
  private Throwable exceptionCausedStop = null;

  @Inject
  private RuntimeClock(
      final Timer timer,
      @Parameter(Clock.StartHandler.class)
          final InjectionFuture<Set<EventHandler<StartTime>>> startHandler,
      @Parameter(Clock.StopHandler.class)
          final InjectionFuture<Set<EventHandler<StopTime>>> stopHandler,
      @Parameter(Clock.RuntimeStartHandler.class)
          final InjectionFuture<Set<EventHandler<RuntimeStart>>> runtimeStartHandler,
      @Parameter(Clock.RuntimeStopHandler.class)
          final InjectionFuture<Set<EventHandler<RuntimeStop>>> runtimeStopHandler,
      @Parameter(Clock.IdleHandler.class)
          final InjectionFuture<Set<EventHandler<IdleClock>>> idleHandler) {

    this.timer = timer;
    this.startHandler = startHandler;
    this.stopHandler = stopHandler;
    this.runtimeStartHandler = runtimeStartHandler;
    this.runtimeStopHandler = runtimeStopHandler;
    this.idleHandler = idleHandler;

    LOG.log(Level.FINE, "RuntimeClock instantiated.");
  }

  /**
   * Schedule a new Alarm event in `offset` milliseconds into the future,
   * and supply an event handler to be called at that time.
   * @param offset Number of milliseconds into the future relative to current time.
   * @param handler Event handler to be invoked.
   * @return Newly scheduled alarm.
   * @throws IllegalStateException if the clock is already closed.
   */
  @Override
  public Time scheduleAlarm(final int offset, final EventHandler<Alarm> handler) {

    final Time alarm = new ClientAlarm(this.timer.getCurrent() + offset, handler);

    if (LOG.isLoggable(Level.FINEST)) {

      final int eventQueueLen;
      synchronized (this.schedule) {
        eventQueueLen = this.numClientAlarms;
      }

      LOG.log(Level.FINEST,
          "Schedule alarm: {0} Outstanding client alarms: {1}",
          new Object[] {alarm, eventQueueLen});
    }

    synchronized (this.schedule) {

      if (this.isClosed) {
        throw new IllegalStateException("Scheduling alarm on a closed clock");
      }

      if (alarm.getTimestamp() > this.lastClientAlarm) {
        this.lastClientAlarm = alarm.getTimestamp();
      }

      assert this.numClientAlarms >= 0;
      ++this.numClientAlarms;

      this.schedule.add(alarm);
      this.schedule.notify();
    }

    return alarm;
  }

  /**
   * Stop the clock. Remove all other events from the schedule and fire StopTimer
   * event immediately. It is recommended to use close() method for graceful shutdown
   * instead of stop().
   */
  @Override
  public void stop() {
    this.stop(null);
  }

  /**
   * Stop the clock on exception.
   * Remove all other events from the schedule and fire StopTimer event immediately.
   * @param exception Exception that is the cause for the stop. Can be null.
   */
  @Override
  public void stop(final Throwable exception) {

    LOG.entering(CLASS_NAME, "stop");

    synchronized (this.schedule) {

      if (this.isClosed) {
        LOG.log(Level.FINEST, "Clock has already been closed");
        return;
      }

      this.isClosed = true;
      this.exceptionCausedStop = exception;

      final Time stopEvent = new StopTime(this.timer.getCurrent());
      LOG.log(Level.FINE,
          "Stop scheduled immediately: {0} Outstanding client alarms: {1}",
          new Object[] {stopEvent, this.numClientAlarms});

      assert this.numClientAlarms >= 0;
      this.numClientAlarms = 0;

      this.schedule.clear();
      this.schedule.add(stopEvent);
      this.schedule.notify();
    }

    LOG.exiting(CLASS_NAME, "stop");
  }

  /**
   * Wait for all client alarms to finish executing and gracefully shutdown the clock.
   */
  @Override
  public void close() {

    LOG.entering(CLASS_NAME, "close");

    synchronized (this.schedule) {

      if (this.isClosed) {
        LOG.exiting(CLASS_NAME, "close", "Clock has already been closed");
        return;
      }

      this.isClosed = true;

      final Time stopEvent = new StopTime(Math.max(this.timer.getCurrent(), this.lastClientAlarm + 1));
      LOG.log(Level.FINE,
          "Graceful shutdown scheduled: {0} Outstanding client alarms: {1}",
          new Object[] {stopEvent, this.numClientAlarms});

      this.schedule.add(stopEvent);
      this.schedule.notify();
    }

    LOG.exiting(CLASS_NAME, "close");
  }

  /**
   * Check if there are no client alarms scheduled.
   * @return True if there are no client alarms in the schedule, false otherwise.
   */
  @Override
  public boolean isIdle() {
    synchronized (this.schedule) {
      assert this.numClientAlarms >= 0;
      return this.numClientAlarms == 0;
    }
  }

  /**
   * The clock is closed after a call to stop() or close().
   * A closed clock cannot add new alarms to the schedule, but, in case of the
   * graceful shutdown, can still invoke previously scheduled ones.
   * @return true if closed, false otherwise.
   */
  @Override
  public boolean isClosed() {
    synchronized (this.schedule) {
      return this.isClosed;
    }
  }

  /**
   * Register event handlers for the given event class.
   * @param eventClass Event type to handle. Must be derived from Time.
   * @param handlers One or many event handlers that can process given event type.
   * @param <T> Event type - must be derived from class Time. (i.e. contain a timestamp).
   */
  @SuppressWarnings("checkstyle:hiddenfield")
  private <T extends Time> void subscribe(final Class<T> eventClass, final Set<EventHandler<T>> handlers) {
    for (final EventHandler<T> handler : handlers) {
      LOG.log(Level.FINEST, "Subscribe: event {0} handler {1}", new Object[] {eventClass.getName(), handler});
      this.handlers.subscribe(eventClass, handler);
    }
  }

  /**
   * Main event loop.
   * Set up the event handlers, and go into event loop that polls the schedule and process events in it.
   */
  @Override
  public void run() {

    LOG.entering(CLASS_NAME, "run");

    try {

      LOG.log(Level.FINE, "Subscribe event handlers");

      subscribe(StartTime.class, this.startHandler.get());
      subscribe(StopTime.class, this.stopHandler.get());
      subscribe(RuntimeStart.class, this.runtimeStartHandler.get());
      subscribe(RuntimeStop.class, this.runtimeStopHandler.get());
      subscribe(IdleClock.class, this.idleHandler.get());

      LOG.log(Level.FINE, "Initiate runtime start");
      this.handlers.onNext(new RuntimeStart(this.timer.getCurrent()));

      LOG.log(Level.FINE, "Initiate start time");
      this.handlers.onNext(new StartTime(this.timer.getCurrent()));

      while (true) {

        LOG.log(Level.FINEST, "Enter clock main loop.");

        try {

          if (this.isIdle()) {
            // Handle an idle clock event, without locking this.schedule
            this.handlers.onNext(new IdleClock(this.timer.getCurrent()));
          }

          final Time event;
          final int eventQueueLen;
          synchronized (this.schedule) {

            while (this.schedule.isEmpty()) {
              this.schedule.wait();
            }

            assert this.schedule.first() != null;

            // Wait until the first scheduled time is ready.
            // NOTE: while waiting, another alarm could be scheduled with a shorter duration
            // so the next time I go around the loop I need to revise my duration.
            while (true) {
              final long waitDuration = this.timer.getDuration(this.schedule.first());
              if (waitDuration <= 0) {
                break;
              }
              this.schedule.wait(waitDuration);
            }

            // Remove the event from the schedule and process it:
            event = this.schedule.pollFirst();

            if (event instanceof ClientAlarm) {
              --this.numClientAlarms;
              assert this.numClientAlarms >= 0;
            }

            eventQueueLen = this.numClientAlarms;
          }

          assert event != null;

          LOG.log(Level.FINER,
              "Process event: {0} Outstanding client alarms: {1}", new Object[] {event, eventQueueLen});

          if (event instanceof Alarm) {
            ((Alarm) event).run();
          } else {
            this.handlers.onNext(event);
            if (event instanceof StopTime) {
              break; // we're done.
            }
          }

        } catch (final InterruptedException expected) {
          LOG.log(Level.FINEST, "Wait interrupted; continue event loop.");
        }
      }

      this.handlers.onNext(new RuntimeStop(this.timer.getCurrent(), this.exceptionCausedStop));

    } catch (final Exception e) {

      LOG.log(Level.SEVERE, "Error in runtime clock", e);
      this.handlers.onNext(new RuntimeStop(this.timer.getCurrent(), e));

    } finally {
      LOG.log(Level.FINE, "Runtime clock exit");
    }

    LOG.exiting(CLASS_NAME, "run");
  }
}
